src/main/java/org/embulk/parser/jsonpath/JsonpathParserPlugin.java

```diff
@@ -86,91 +93,82 @@ public void run(TaskSource taskSource, Schema schema,
-                        continue;
-                    }
+                if (!json.isArray()) {
+                    skipOrThrow(new JsonRecordValidateException(format(Locale.ENGLISH,
+                            "Json string is not representing array value json='%s'", json)), stopOnInvalidRecord);
+                    continue;
+                }
 
-                    logger.debug("recordValue = " + recordValue.toString());
-                    final Map<Value, Value> record = recordValue.asMapValue().map();
+                for (JsonNode recordValue : json) {
                     try {
+                        if (recordValue.getNodeType() != JsonNodeType.OBJECT) {
+                            throw new JsonRecordValidateException(format(Locale.ENGLISH,
+                                    "Json string is not representing map value json='%s'", recordValue));
+                        }
+
                         for (Column column : schema.getColumns()) {
-                            Value v = record.get(getColumnNameValue(column));
-                            visitor.setValue(v);
+                            JsonNode value = null;
+                            if (jsonPathMap.containsKey(column)) {
+                                try {
+                                    value = JsonPath.using(JSON_PATH_CONFIG).parse(recordValue).read(jsonPathMap.get(column));
+                                }
+                                catch (PathNotFoundException e) {
+                                    // pass (value is nullable)
+                                }
+                            } else {
+                                value = recordValue.get(column.getName());
+                            }
+                            visitor.setValue(value);
                             column.visit(visitor);
                         }
+
+                        pageBuilder.addRecord();
                     }
                     catch (DataException e) {
                         skipOrThrow(e, stopOnInvalidRecord);
                         continue;
                     }
-
-                    pageBuilder.addRecord();
                 }
             }
 
             pageBuilder.finish();
         }
     }
 
-    private void setColumnNameValues(Schema schema)
+    private Map<Column, String> createJsonPathMap(PluginTask task, Schema schema)
     {
-        ImmutableMap.Builder<String, Value> builder = ImmutableMap.builder();
-        for (Column column : schema.getColumns()) {
-            String name = column.getName();
-            builder.put(name, newString(name));
+        ImmutableMap.Builder<Column, String> builder = ImmutableMap.builder();
+        for (int i = 0; i < schema.size(); i++) {
+            ColumnConfig config  = task.getSchemaConfig().getColumn(i);
+            if (config.getOption().has("path")) {
+                builder.put(schema.getColumn(i), config.getOption().get(String.class, "path"));
```

Comment:
I think your solution is smarter than current one.

I've defined `JsonpathColumnOption`, instead of call `getOption.get(xxx)`.